src/ZendeskApi.Client/Models/CursorPaginatedIteratorFactory.cs

```diff
@@ -0,0 +1,27 @@
+using System;
+using Microsoft.Extensions.DependencyInjection;
+using ZendeskApi.Client.Responses;
+
+namespace ZendeskApi.Client.Models
+{
+    public interface ICursorPaginatedIteratorFactory
+    {
+        CursorPaginatedIterator<T> Create<T>(ICursorPagination<T> response);
+    }
+
+    public class CursorPaginatedIteratorFactory : ICursorPaginatedIteratorFactory
+    {
+        private static IServiceProvider serviceProvider;
```

Comment:
I think we are getting closer to the end result here. There are still a few points I would like to make...

This `serviceProvider` field member should not be static, it should be `readonly`. We should also not be injecting the `IServiceProvder` into a class. This endangers running into the [Service Locator anti-pattern](https://blog.ploeh.dk/2010/02/03/ServiceLocatorisanAnti-Pattern/). I won't go into too much detail here but in short the IoC container is the thing responsible for choosing the correct dependency - in general you don't want a consuming class acting as a locator. The usual pattern for injecting dependencies in dotnet is closer to the following

```c#
namespace ZendeskApi.Client.Pagination
{
    public interface ICursorPaginatedIteratorFactory
    {
        CursorPaginatedIterator<T> Create<T>(ICursorPagination<T> response);
    }

    public class CursorPaginatedIteratorFactory : ICursorPaginatedIteratorFactory
    {
        // readonly field as only set in constructor
        private readonly IZendeskApiClient zendeskApiClient;

        public CursorPaginatedIteratorFactory(IZendeskApiClient _zendeskApiClient)
        {
            // inject IZendeskApiClient after registering in the container
            zendeskApiClient = _zendeskApiClient;
        }

        public CursorPaginatedIterator<T> Create<T>(ICursorPagination<T> response)
        {
            return new CursorPaginatedIterator<T>(response, zendeskApiClient);
        }
    }
}
```

This means a consumer looks like the following:

```c#
public class YourConsumerClass
{
    private readonly CursorPaginatedIteratorFactory iteratorFactory;

    public YourConsumerClass(ICursorPaginatedIteratorFactory _iteratorFactory)
    {
        iteratorFactory = _iteratorFactory;
    }

    public void YourMethod(ICursorPagination<YourType> response)
    {
        // Use the factory to get the iterator
        CursorPaginatedIterator<YourType> iterator = iteratorFactory.Create(response);

        // Rest of your logic using the iterator...
    }
}

```

To summarise, from where you are we may consider

* alter `ICursorPaginatedIteratorFactory`:
  * inject `IZendeskApiClient`  directly, instead of the `IServiceProvider`
  * make the `IZendeskApiClient` field member private, readonly - not static
* Register your new interface/implementation in the IoC container. We have helper methods which we recommend using [here](https://github.com/justeat/ZendeskApiClient/blob/621cde89b7b888e0ca11cce38d79e442aa0a9e8a/src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs#L9C7-L9C7).


### Aside

I want to reiterate that, in an ideal world, consumers would be able to iterate over the response from paginated endpoints **without** injecting another dependency on top of `IZendeskClient`. One approach we already discussed was to change the return type of all CBP endpoints to be the iterator itself. We discounted this because of the blast-radius of the change. It's worth noting that this is probably not that dangerous. As you have explored recently these existing CBP methods don't work properly so people are probably not using it meaningfully. So now is probably the best time to make that change. I do understand your hesitancy though and it may be one for us owners to pick up if you wanted to.

Another approach that works around this is to use the adapter pattern. At the moment the CBP endpoints return `ICursorPagination<T>`. An adapter class would adapt the adapt the method signature of the CBP methods to return `CursorPaginatedIterator<T>`. It would work in a similar way to the IteratorFactory you have but the usage is different. You would expose the adapted resources on the IZendeskClient interface so users would consume like so:

```c#
client.AdaptedTicketAudit.GetAllAsync()
``` 
 where the `AdaptedTicketAudit` is named for clarity but essentially the underlying resource would implement a different interface to `ITicketAuditResource`

Not suggesting this adapter as an approach because the amount of code changes required here is massive and I prefer the formerly suggested. But it just came to my mind so wanted to share in case it sparked anything.

Reply:
thanks @mikerogers123, I believe I made the (small) changes you requested 😅 
to be very clear, I'm a Ruby developer who picked up this work of updating the C# SDK in order to unblock Zendesk from rolling out the OBP deprecation. How is it looking after this last commit? The usage I wrote in the PR description remains unchanged and I'm registering the Factory [here](https://github.com/justeat/ZendeskApiClient/pull/322/files#diff-93228efc1744d0ed92be5a48775f5aaf79b820a7a4226c38d096c996bab407c0R38)

Reply:
Thanks for that @fbvilela the changes look good overall. Have added a few more minor comments to sort, then we should be close to closing this one